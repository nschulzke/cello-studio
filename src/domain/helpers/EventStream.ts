export const EventStream = () => {
  let subscribers = [];

  return {
    subscribe: (callback) => {
      subscribers.push(callback);
      return () => { subscribers = subscribers.filter(sub => sub !== callback) }
    },
    publish: (event) => {
      subscribers.forEach(sub => sub(event));
    }
  }
}
