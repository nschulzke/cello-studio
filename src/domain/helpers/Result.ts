interface Success<T> {
  success: true;
  data: T;
}

interface Failure {
  success: false;
  data: string;
}

export type Result<T> = Success<T> | Failure;

export const Success = <T>(data: T): Result<T> => ({
  success: true,
  data,
});

export const Failure = (message: string): Result<any> => ({
  success: false,
  data: message,
});

export const fromBool = (bool: boolean, failureMessage: string): Result<boolean> => (
  bool ? { success: true, data: true } : { success: false, data: failureMessage }
)

export const pipe = (...tests: Array<(value: any) => Result<any>>): Result<any> => (
  tests.reduce((previous, next) => {
    if (previous.success) {
      return next(previous.data);
    } else {
      return previous;
    }
  }, Success(true))
);

export const all = <T>(argument: T, ...tests: Array<(value: Partial<T>) => Result<any>>) => {
  let result = Success(true);
  for (let test of tests) {
    result = test(argument);
    if (!result.success) {
      return result;
    }
  }
  return result;
}

export const lift = <T>(func: (arg: T) => boolean, failureMessage: string) => (arg: T) => (
  fromBool(func(arg), failureMessage)
);
