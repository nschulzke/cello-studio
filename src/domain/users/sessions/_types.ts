export enum Permissions {
  ADMIN = 'ADMIN',
  STUDENT = 'STUDENT',
  NONE = 'NONE',
}

export interface CredentialsRaw {
  email: string;
  password: string;
}

export interface CredentialsHashed {
  email: string;
  hash: string;
}
