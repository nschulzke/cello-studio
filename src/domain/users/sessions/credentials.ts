import * as bcrypt from 'bcrypt';

export const credentials = {
  hash: ({ email, password }) => ({
    email,
    hash: bcrypt.hashSync(password, 10)
  }),

  compare: ({ email: sentEmail, password }, { email: trueEmail, hash } = { email: false, hash: '' }) =>
    sentEmail === trueEmail
    && bcrypt.compareSync(password, hash),
}
