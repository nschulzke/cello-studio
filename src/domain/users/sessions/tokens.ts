import * as jwt from 'jsonwebtoken';

export const prepareTokens = ({ secret, duration }) => ({
  encode: (payload) =>
    jwt.sign({ data: payload }, secret, { expiresIn: duration }),

  decode: (token) => {
    try {
      return jwt.verify(token, secret).data;
    } catch (e) {
      return false;
    }
  },
})
