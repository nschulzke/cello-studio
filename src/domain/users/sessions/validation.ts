import { fromBool } from 'domain/helpers/Result';

var emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export const emailIsValid = ({ email }) =>
  fromBool(emailRegex.test(String(email)), 'Email is invalid.');
