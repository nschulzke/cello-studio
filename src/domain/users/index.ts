import * as Result from 'domain/helpers/Result';
import { profiles } from './profiles';
import { prepareUsers, prepareTokens, credentials } from './sessions';

export const hydrateUsers = (stream, state, secret) => {
  const tokens = prepareTokens({ secret, duration: '1m' });
  const users = prepareUsers({ credentials, tokens });

  return {
    register({ email, password }) {
      let result = users.register(state, { email, password });
      if (result.success) {
        state = result.data;
        stream.publish(state);
        return Result.Success(tokens.encode(email));
      } else {
        return result;
      }
    },

    login({ email, password }) {
      return users.login(state, { email, password })
    },

    verify(token) {
      return users.verify(state, token);
    },

    getProfile(email) {
      if (typeof state[email] === 'object') {
        return Result.Success(profiles.get(state, email));
      } else {
        return Result.Failure('No such email address.');
      }
    },

    updateProfile(email, profile) {
      if (typeof state[email] === 'object') {
        state = profiles.update(state, email, profile);
        stream.publish(state);
        return Result.Success(state[email]);
      } else {
        return Result.Failure('No such email address.');
      }
    }
  }
}
