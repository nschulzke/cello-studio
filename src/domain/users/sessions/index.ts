import * as Result from 'domain/helpers/Result';
import { emailIsValid } from './validation';
import { Permissions } from './_types';

const emailIsFree = ({ userIndex, email, password }) =>
  userIndex[email] === undefined || userIndex[email].hash === undefined
    ? Result.Success(true)
    : Result.Failure('Email is in use.');

export const prepareUsers = ({ credentials, tokens }) => ({
  register: (userIndex, { email, password }) => (
    Result.all({ userIndex, email, password },
      emailIsFree,
      emailIsValid,
      ({ userIndex, email, password }) => (Result.Success({
        ...userIndex,
        [email]: { ...userIndex[email], ...credentials.hash({ email, password }), permissions: Permissions.STUDENT }
      }))
    )
  ),

  login: (userIndex, { email, password }) => {
    return credentials.compare({ email, password }, userIndex[email])
      ? Result.Success({ email, token: tokens.encode(email), permissions: userIndex[email].permissions })
      : Result.Failure('Invalid email or password.');
  },

  fetchUser: (userIndex, token) => {
    let result = tokens.decode(token);
    if (result === false) {
      return Result.Failure('Your token could not be validated.');
    } else {
      return Result.Success(userIndex[result]);
    }
  },

  permissions: (userIndex, token) => {
    let email = tokens.decode(token);
    if (email && userIndex[email] && userIndex[email].permissions) {
      return userIndex[email].permissions;
    } else {
      return Permissions.NONE;
    }
  }
});

export { prepareTokens } from './tokens';
export { credentials } from './credentials';
