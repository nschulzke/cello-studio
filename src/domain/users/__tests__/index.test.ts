import { hydrateUsers } from '../index';
import { prepareTokens } from '../sessions/tokens';
import { credentials } from '../sessions/credentials';
import { EventStream } from 'domain/helpers/EventStream';
import { Result } from 'domain/helpers/Result';
import { Permissions } from 'domain/types';

describe('index', () => {
  const secret = 'testpass';
  let newUser = { email: 'test@example.com', password: 'mynewpass' };
  let userModel = {};
  let fullUserModel = { [newUser.email]: { ...credentials.hash(newUser), name: 'Test', parentName: 'Test2', permissions: Permissions.STUDENT } };
  let stream = EventStream();

  describe('register', () => {
    it('registers a new user successfully', () => {
      let users = hydrateUsers(stream, userModel, secret);
      let result = users.register(newUser);
      expect(result.success).toBe(true);
      let token = result.data;
      expect(typeof token).toBe('string');
    });

    it('sends events to the stream', async () => {
      function checkStream(): Promise<Result<any>> {
        return new Promise((resolve) => {
          let unsub = stream.subscribe((event) => {
            resolve(event);
          });
          let users = hydrateUsers(stream, userModel, secret);
          users.register(newUser);
          unsub();
        });
      }

      let result = await checkStream();
      expect(result[newUser.email]).toBeTruthy();
    });

    it('rejects if the user is already registered', () => {
      let users = hydrateUsers(stream, fullUserModel, secret);
      let failure = users.register(newUser);
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });

    it('rejects an invalid email address', () => {
      let users = hydrateUsers(stream, fullUserModel, secret);
      let failure = users.register({ email: 'email', password: 'testing' });
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });
  });

  describe('login', () => {
    it('accepts correct login information', () => {
      let users = hydrateUsers(stream, fullUserModel, secret);
      let result = users.login(newUser)
      expect(result.success).toBe(true);
      expect(typeof result.data).toBe('object');
      expect(typeof result.data.token).toBe('string');
      expect(result.data.email).toEqual(newUser.email);
      expect(result.data.permissions).toEqual(Permissions.STUDENT);
    });

    it('rejects a bad email address', () => {
      let users = hydrateUsers(stream, fullUserModel, secret);
      let failure = users.login({ ...newUser, email: 'hi!' });
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });

    it('rejects a bad email password', () => {
      let users = hydrateUsers(stream, fullUserModel, secret);
      let failure = users.login({ ...newUser, password: 'hi!' });
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });
  });

  describe('verify', () => {
    let tokens = prepareTokens({ secret, duration: '1h' });
    let users = hydrateUsers(stream, fullUserModel, secret);
    let token = tokens.encode(newUser.email);

    it('returns the user object', () => {
      expect(users.fetchUser(token).data).toEqual(fullUserModel[newUser.email]);
    });

    it('rejects invalid tokens', () => {
      let failure = users.fetchUser('badtoken');
      expect(failure.success).toBe(false);
      expect(failure.data).toMatchSnapshot();
    });
  });

  describe('profiles', () => {
    describe('get', () => {
      it('gets a student profile from the index', () => {
        let users = hydrateUsers(stream, fullUserModel, secret);
        let got = users.getProfile(newUser.email);
        expect((got as any).hash).toEqual(undefined);
        expect(got.success).toBe(true);
        expect(got.data.name).toEqual('Test');
        expect(got.data.parentName).toEqual('Test2');
        expect(got.data.contactEmail).toEqual('');
      });

      it('returns a failure if the email is bad', () => {
        let users = hydrateUsers(stream, fullUserModel, secret);
        let got = users.getProfile('bademail');
        expect(got.success).toBe(false);
      });
    });

    describe('update', () => {
      it('updates a student profile', () => {
        let users = hydrateUsers(stream, fullUserModel, secret);
        let newContactEmail = 'test@example.com';
        let result = users.updateProfile(newUser.email, { contactEmail: newContactEmail });
        expect(result.success).toBe(true);
        expect(result.data.contactEmail).toEqual(newContactEmail);
        expect(users.getProfile(newUser.email).data.contactEmail).toEqual(newContactEmail);
      });

      it('returns a failure if the email is bad', () => {
        let users = hydrateUsers(stream, fullUserModel, secret);
        let got = users.updateProfile('bademail', {});
        expect(got.success).toBe(false);
      });
    });
  });
});
