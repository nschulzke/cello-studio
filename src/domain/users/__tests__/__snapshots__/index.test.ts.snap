// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index login rejects a bad email address 1`] = `"Invalid email or password."`;

exports[`index login rejects a bad email password 1`] = `"Invalid email or password."`;

exports[`index register rejects an invalid email address 1`] = `"Email is invalid."`;

exports[`index register rejects if the user is already registered 1`] = `"Email is in use."`;

exports[`index verify rejects invalid tokens 1`] = `"Your token could not be validated."`;
